### DISCUSSION TITLE
Unit tests

### DISCUSSION THREAD

**Discussion by fzs** [troglobit/mdnsd#54] [General]
Have you ever considered adding unit tests to the project?
I am reading the code and since I do not have the most in-depth mDNS knowledge I have a hard time understanding what some parts of the code do. So aside from helping with refactoring the code, unit tests could also be helpful to understand what certain code is supposed to do.

**Comment by troglobit** [comment]
Yes, I've planned it ever sine I had a breakthrough in some of my other projects using `unshare(1)`.  Let me see what I can whip up later today, shouldn't be too much work to get some basics up and running :)

Again, sorry for missing all this (and the PRs). Seems my project notifications were out of whack on GitHub for some reason.

**Reply by troglobit** [reply]
Was more work, will try to finish framework and initial test(s) this weekend instead.

**Reply by fzs** [reply]
Well, I would think this is quite some work and not that quickly whipped up. 
What unit test framework do you usually use?

**Reply by troglobit** [reply]
Well I was more going for a "component" test suite, not a unit test, i.e., running mdnsd with various configs, parsing the output from tshark/tcpdump, and probing it with mquery.  This has worked well in other projects of mine.

**Reply by troglobit** [reply]
Just pushed the first iteration, very simple framework and only a single test so far, but it hopefully gives you an idea of what I was going for.

See the commit message for d77403799a1ec3db388db50e96b21899604ee7c4 for a quick "how to".

**Reply by fzs** [reply]
How come that running `make check` as a user works, but running `discover.sh` as a user doesn't work and needs to be run with `sudo`?

**Reply by troglobit** [reply]
Because it creates network namespaces to emulate isolated devices.  You can call the test like this to run it stand-alone:

    $ unshare -mrun test/discover.sh

or

    $ make check TESTS=discover.sh

**Reply by fzs** [reply]
Neither seems to work. The former complains about not finding mdnsd and the latter doesn't run any tests at all

**Reply by troglobit** [reply]
Sorry, didn't verify on my machine.

This works for me on Linux Mint (based on Ubuntu):

    cd test
    unshare -mrun ./discover.sh cd

and also the following, from the top directory of the project:

    make check TESTS=discover.sh

In the first variant you see the logs directly on stdout/stderr, and in the latter logs are saved in `discover.log`.

**Reply by fzs** [reply]
I am running this on a Linux Mint 20.3 VM.

```
$ make check TESTS=discover.sh
Making check in examples
make[1]: Entering directory '/home/florian/Devel/External/mdnsd/examples'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/florian/Devel/External/mdnsd/examples'
Making check in libmdnsd
make[1]: Entering directory '/home/florian/Devel/External/mdnsd/libmdnsd'
  CC       libmdnsd_la-mdnsd.lo
  CCLD     libmdnsd.la
make[1]: Leaving directory '/home/florian/Devel/External/mdnsd/libmdnsd'
Making check in man
make[1]: Entering directory '/home/florian/Devel/External/mdnsd/man'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/florian/Devel/External/mdnsd/man'
Making check in src
make[1]: Entering directory '/home/florian/Devel/External/mdnsd/src'
  CCLD     mquery
  CCLD     mdnsd
make[1]: Leaving directory '/home/florian/Devel/External/mdnsd/src'
Making check in test
make[1]: Entering directory '/home/florian/Devel/External/mdnsd/test'
Making check in src
make[2]: Entering directory '/home/florian/Devel/External/mdnsd/test/src'
make  check-TESTS
make[3]: Entering directory '/home/florian/Devel/External/mdnsd/test/src'
make[4]: Entering directory '/home/florian/Devel/External/mdnsd/test/src'
make[5]: Entering directory '/home/florian/Devel/External/mdnsd/test/src'
make[5]: Nothing to be done for 'discover.sh.log'.
make[5]: Leaving directory '/home/florian/Devel/External/mdnsd/test/src'
fatal: making test-suite.log: failed to create discover.sh.trs
fatal: making test-suite.log: failed to create discover.sh.log
make[4]: *** [Makefile:698: test-suite.log] Error 1
make[4]: Leaving directory '/home/florian/Devel/External/mdnsd/test/src'
make[3]: *** [Makefile:807: check-TESTS] Error 2
make[3]: Leaving directory '/home/florian/Devel/External/mdnsd/test/src'
make[2]: *** [Makefile:887: check-am] Error 2
make[2]: Leaving directory '/home/florian/Devel/External/mdnsd/test/src'
make[1]: *** [Makefile:562: check-recursive] Error 1
make[1]: Leaving directory '/home/florian/Devel/External/mdnsd/test'
make: *** [Makefile:518: check-recursive] Error 1
```

I tried `unshare -mrun ./iprecords.sh` in the `test` folder again and that worked for me. So that is fine for me to be able to just test my test shell scripts. But the Make way seems not entirely reliable for the general public, it looks like.

**Reply by troglobit** [reply]
Thanks for the log, that looks really strange. It's the same setup I use in many other project and have never seen anything like this ... maybe I forgot to set up something?  I'll check later with a clean install without any dev packages installed, hopefully I can reproduce.

Just one thing, did you re-run `./autogen.sh` after I pushed the `test/` directory?

Glad you can run with unshare directly, for now at least.

**Reply by fzs** [reply]
> Just one thing, did you re-run ./autogen.sh after I pushed the test/ directory?

I don't know. I did now, made clean and ran `make check TESTS=discover.sh` again. Same erroneous result.

**Reply by troglobit** [reply]
Thanks for trying it out, now we know.  I'll try to get to the bottom of this as soon as possible.  Appreciate you taking the time to report and follow-up on this! <3

**Reply by troglobit** [reply]
Tested this out on three different computers today: Ubuntu 22.04, Linux Mint 21, and Linux Mint 20.2. Neither of them had all the required packages installed, all had `build-essential` though, so both `gcc` and `make` were there.  The magic commands that made `make check` work was:

```sh
cd mdnsd/
sudo apt install build-essential autoconf automake libtool pkg-config \
                 libcmocka-dev autotools-dev libsigsgv2 m4 libltdl-dev
./autogen.sh
./configure
make -j9 check
```

I could not reproduce your problem at all, unfortunately.

**Reply by fzs** [reply]
Well, it stays the same for me with the recent distcheck changes. I just cloned the current repo to a different, relatively untouched Mint 20.1 VM. All of your above commands produce the expected result. `make check TESTS=addr` works, but `make check TESTS=discover.sh` does not. I don't know much about automake, my suspicion lies on the '.sh' ending, because of the files a working `make check` produces and the log messages in the error case.

**Reply by troglobit** [reply]
Ah ... that's probably it. We're mixing apples and oranges -- tests both in C and .sh, with and without ending.  Good find!  I had a feeling this was going to bite us.  Maybe we can play around with dropping the .sh from shell scripts (and the `TEST_EXTENSIONS`)?  At least to test the theory.

**Reply by fzs** [reply]
I can try that a bit later. But why does it still work for you but not on my systems?

**Reply by troglobit** [reply]
Good question, and I don't know what I was smoking earlier (or what state my clone was in), but I no longer get `make check TESTS=discover.sh` to work either -- so that's progress I guess, we now both have the same issue :grin:

**Reply by fzs** [reply]
I played around with removing TEST_EXTENSIONS and renaming the scripts but could not get it to work.

**Reply by troglobit** [reply]
Thank you for giving it a go!  I just created issue #66 for this.

**Comment by fzs** [comment]
Have you ever considered actual unit tests?
I am intrigued by your setup with unshare, I will have to learn more about it and an learn to use it.
But I also see value in actual unit tests, while the current ones are more integration tests. They require a Linux system and the `mdnsd` and `mquery` binaries (and are also a bit testing my code with my code). We integrate mdnsd into a non-linux embedded OS, so it is not a 1-to-1 mapping. We are mostly using [cmocka ](https://cmocka.org/)for unit tests, and currently have tests for mdnsd external to the source tree. I wonder if it would be of value and interest to integrate them with the project. It does add a dependency on the test suite library, though.
I also sometimes find it easier to test with smaller unit tests during development, like for example when implementing complexer checks for link-local, site-local. global addresses replacing each other.

**Reply by troglobit** [reply]
Adding actual unit tests to an existing program/library is a LOT of work.  I've considered it in the past but I have not had much time to spend on this low-prio project lately. 

You are much welcome to add unit tests with cmocka if you feel up to it!  I can help out with adding configure flags to run only unit tests and disable component tests, if you like?  I do prefer having it all in the same test/ sub-directory so a regular build with "make check" runs everything.

**Reply by fzs** [reply]
I would need some help adding the dependency on the cmocka library to the Makefiles, as I have no experience with autotools.
I would suggest adding a `src` directory under `test` and put test source code files there.

**Reply by troglobit** [reply]
I've just pushed initial support, including a bare bones example, to master in cda3cb4.  Hope it helps :)

Not sure if the src subdirectory should hold run the tests, or if we should run them together from the parent test directory.  You try it out and tell me what you think.

**Reply by fzs** [reply]
Excellent!
Two questions: 
When testing `addr.c` or `conf.c`, how can I link these into the test binary?
Is there a way to add linker flags (`-Wl --wrap=something`) for certain `noinst_PROGRAMS`?

I should add that I toyed around with `LDADD` and `LDFLAGS` but I have no idea if this is the correct way to do it.

```
addr_LDADD   = ../../src/addr.o $(LIBOBJS)
addr_LDFLAGS = $(AM_LDFLAGS) -Wl,--wrap=getifaddrs
```

**Reply by troglobit** [reply]
Yup, that's the correct way to do it.  You can also set it globally for all test binaries with the AM_foo variables.  And if you want test specific flags you need to do it like you did in your example, because the AM_foo variables are not included if there are test specific ones declared.

**Comment by troglobit** [comment]
@fzs heads-up, I've just pushed a few changes to the `test/` directory to address issues with `make distcheck`.  Not sure if that fixes your above issues with `make check` as well ...
